server: Complete balance update when SubmitSegment() returns

yondonfu opened the pull request:
**What does this pull request do? Explain your changes. (required)**
<!-- A clear and concise description of what this pull request does. -->

This PR implements balance update completion within `SubmitSegment()` with logic that differs depending on the update's status at the time of completion. There are three possible update statuses:

- `Staged`: This is the initial status of the `BalanceUpdate` type. The update remains in this status until the HTTP request for a segment (and its payment header) is submitted.
- `CreditSpent`: This is the status of the `BalanceUpdate` after the HTTP request for a segment is submitted and before a response from O is received. Note: there is no distinction between a HTTP request with a success code and a failure code - as long as the HTTP request was submitted without an error in the Go code, the `BalanceUpdate` status is `CreditSpent`
- `ReceivedChange`: This is the status of the `BalanceUpdate` after a response with transcoded results is received from O

**Specific updates (required)**
<!--- List out all significant updates your code introduces -->
- Add the `completeBalanceUpdate()` helper in `server/segment_rpc.go` which executes the logic for completing a balance update depending on the update's status
- Defer the execution of `completeBalanceUpdate()` in `SubmitSegment()` to guarantee its execution before the function returns
- Change the status of `BalanceUpdate` created when preparing a payment at different points of the execution of `SubmitSegment`

**How did you test each of these updates (required)**
<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->

Added unit tests.

**Does this pull request close any open issues?**
<!-- Fixes # -->

Fixes #898 

**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

yondonfu committed:
server: Complete balance update when SubmitSegment() returns

kyriediculous commented on server/segment_rpc.go:
```diff
@@ -386,9 +384,19 @@ func SubmitSegment(sess *BroadcastSession, seg *stream.HLSSegment, nonce uint64)
-	// Set balUpdate.Status = ReceivedChange
+	// We treat a response as "receiving change" where the change is the difference between the credit and debit for the update
+	balUpdate.Status = ReceivedChange
+	priceInfo := sess.OrchestratorInfo.PriceInfo
```
Since we're fetching the priceInfo from the session after calling `updateOrchestratorInfo`, we'll be overwriting any old PriceInfo on which B's expected price was based and we're using B's expected price (if acceptable) to Debit the balance from the orchestrator side. 

This could result in a discrepancy between the actual debit amounts. 

https://github.com/livepeer/go-livepeer/blob/40c37f7264f4f8ac980cd24e17d15b3a98571fd9/server/segment_rpc.go#L152

yondonfu replied:
Ah good catch. I think we can fix this by deferring execution of `updateOrchestratorInfo()` until the very end up of `SubmitSegment()` since the only current requirement is that the session's `OrchestratorInfo` is updated when `SubmitSegment()` returns

yondonfu replied:
Fixed in 9979bf4

kyriediculous commented on server/segment_rpc.go:
```diff
@@ -386,9 +384,19 @@ func SubmitSegment(sess *BroadcastSession, seg *stream.HLSSegment, nonce uint64)
+	// We treat a response as "receiving change" where the change is the difference between the credit and debit for the update
+	balUpdate.Status = ReceivedChange
+	priceInfo := sess.OrchestratorInfo.PriceInfo
+	if priceInfo != nil {
```
I don't think priceInfo can ever be `nil` here, can it? 

1. It would mean expected price is nil , which would trigger an update from the orchestrator and we're using the value after update here (which I don't think we should, see previous comment).
2. it's not possible for O to set a nil priceInfo on node startup, meaning when O is picked up from orchestratorPool and its info is requested it is never nil on the session?

yondonfu replied:
`priceInfo` would be `nil` when the node is running in off-chain mode because Os don't include a price in the `OrchestratorInfo` message in off-chain mode.

kyriediculous replied:
Right, I tend to always forget about offchain mode existing😅 . Thanks !

kyriediculous commented on server/segment_rpc.go:
```diff
@@ -477,31 +485,31 @@ func newBalanceUpdate(sess *BroadcastSession) (*BalanceUpdate, error) {
+
+	// If the update did not include a processed debit then no change was received
+	// so we exit without updating the balance because the credit was spent already
+	if update.Status != ReceivedChange {
```
Doesn't this essentially zero the entire balance ? Why is there no residual balance possible at this point? Thanks.

yondonfu replied:
Yeah this would zero the entire balance. The motivation behind this is that if the segment is submitted (i.e. the `update.Status = CreditSpent`), but B does not receive a successful response before `SubmitSegment()` errors out then B does not know with certainty whether O successfully received the payment and applied credit/debit to its view of B's balance. If we instead add `existingCredit` back to B's balance then it is possible that O did receive the payment and applied credit/debit such that its view of B's balance is `< existingCredit`. Then, if B reuses the session with O, B would receive an insufficient balance error from O because B generates a payment based on an out of sync view of its balance. Considering the existing credit as "spent" in this scenario makes the workflow between B & O less dependent on the two parties having the same exact view of B's balance.

I actually just realized that technically we already zero the balance when `SubmitSegment()` encounters an error because we call [removeSession()](https://github.com/livepeer/go-livepeer/blob/master/server/broadcast.go#L92) which will clear the balance for the session. However, `removeSession()` will actually delete the manifestID key in the underlying `core.Balances` type. Meanwhile, if `update.Status = CreditSpent` when we complete the balance update in `SubmitSegment()`, the balance stored for the manifestID in the underlying `core.Balances` type will be left as `big.NewRat(0, 1)` (and then the entire map entry would be deleted when we call `removeSession()`). I'm ok with leaving this behavior as-is since `removeSession()` actually deletes the keys from `core.Balances`

yondonfu committed:
server: Defer updateOrchestratorInfo() in SubmitSegment()

- Ensure that price for session is only updated after the last known
price is used to compute the fee for results
- Add a test that fails if the behavior described above is not exhibited

kyriediculous commented:
Changes look good to me. 
Needs a rebase !

yondonfu commented:
Rebased - decided to retain two distinct commits for this PR since I think having the second commit as a standalone change might be potentially useful to have in the history (i.e. this change ensures that price and other O info updates for a session happen at the very end of `SubmitSegment()`). I did fix a typo in the commit message though.

kyriediculous reviewed (approved):
That sounds reasonable to me, Good to go 🚀 . 
The gaspricemonitor polling test is flaky again on circle though. 
Can we re-run it for good measure before merging?
